include/painless/parameter.h

```diff
@@ -31,38 +88,66 @@ inline std::string get_base_path() {
 namespace parser {
 
 template <typename T>
-inline bool from_string(const std::string& input, T& value) {
+inline bool from_string(const std::string&, T&, std::false_type) {
+  static_assert(detail::has_istream_operator<T&>::value,
+                "Type does not support the >> istream operator");
+  return false;
+}
+
+template <typename T>
+inline bool from_string(const std::string& input, T& value, std::true_type) {
   std::istringstream ss(input);
   return static_cast<bool>(ss >> value);
 }
 
-inline bool from_string(const std::string& input, bool& value) {
+inline bool from_string(const std::string& input, bool& value, std::true_type) {
   std::istringstream ss(input);
   return static_cast<bool>(ss >> std::boolalpha >> value);
 }
 
-inline bool from_string(const std::string& input, std::string& value) {
+inline bool from_string(const std::string& input,
+                        std::string& value,
+                        std::true_type) {
   value = input;
   return true;
 }
 
+template <typename T>
+inline bool from_string(const std::string& input, T& value) {
+  return from_string(input, value,
+                     typename detail::has_istream_operator<T&>::type{});
+}
+
 }  // namespace parser
 
 namespace printer {
 
 template <typename T>
-inline std::string to_string(const T& value) {
+inline std::string to_string(const T&, std::false_type) {
+  static_assert(detail::has_istream_operator<const T&>::value,
```

Comment:
This should be `has_ostream_operator<…>`.

The `to_string`/`from_string` methods are internal implementation details, while the `Parameter<T>` class is part of the public API.

I think I would therefore prefer the simpler solution where we add two separate `static_assert`s for both `has_ostream_operator` and `has_istream_operator` to the `Parameter` class template. Having an instance of `Parameter<T>` for `T` without any of these operators is not sensible/possible/useful, so we are not artificially restricting the usage.